torch/onnx/_internal/diagnostics/infra/formatter.py

```diff
@@ -19,6 +20,35 @@
 ]
 
 
+class LazyString:
```

Comment:
For my knowledge, if Diagnostic wouldn't be enabled by default, do we still need lazy evaluation? Could we hide the "laziness" within the implementation so we don't have to expose and delegate to user when they should use it when adding more diagnostics on their own?

Reply:
Diagnostic is enabled by default. 

And on the contrary to your question, we need lazy evaluation more if it is not enabled by default, because that means we have more places where we don't want to evaluate messages since we don't log it.

There is no way around it for delayed evaluation. User must pass a function or lambda w/ args. I took this design since it is already there in PT2 logging so we are keeping similar log api surface.

There are no checks for violation. I'm not aware of any simple solution. The only thing I can think of is creating a linter, hence this is low priority.